### DISCUSSION TITLE
Embedding parameter validations in classes/models for reuse

### DISCUSSION THREAD

**Discussion by kauedg** [fastapi/fastapi#8807] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Path
app = FastAPI()

class Username(str):
    def __init__(self, arg):
        super(Username, self).__init__()
    

@app.post(
    path="/user/{username}")
def create_user(
        *,
        username: str = Path(..., min_length=3, max_length=8)
    ):
    return { 'username': username }
```

### Description

I want to reuse this `Path` validation code
`username: str = Path(..., min_length=3, max_length=8)`

multiple times in different endpoints. I'm aware that...
```
When you import Query, Path and others from fastapi, they are actually functions.
That when called, return instances of classes of the same name.
```
and that I could declare and use a function like this
```
def username_validate_path():
    return Path(..., min_length=3, max_length=8)
    
@app.post(
    path="/user/{username}")
def create_user(
        *,
        username: str = username_validate_path()
    ):
    return { 'username': username }
```
but that's just ghastly. In order to keep things tidy I'd like to embed the validation in the `Username` class, so I wrote something like this:
```
class Username(str):
    def __init__(self, arg):
        super(Username, self).__init__()

    @classmethod
    def validate_path(cls):
        return Path(
                ...,
                title="Username",
                min_length=3, 
                max_length=8
                example='chewie'
                )

@app.post(
    path="/user/{username}")
def create_user(
        *,
        username: Username = Username.validate_path()
    ):
    return { 'username': username }
```

But I'm still not quite happy with it. What I'd really like to have is something like this...
```
@app.post(
    path="/user/{username}")
def create_user(
        *,
        username: Username, [or]
        username: Username = Username.validate(),
    ):
    return { 'username': username }
```
... and some function or/and decorator would deal with the type of the parameter and it's validation.

Is it possible? 

### Operating System

Linux

### FastAPI Version

0.63.0

### Python Version

Python 3.8.10

**Comment by be-wise-be-kind** [comment]
An approach that I've found not to be too awful is to use a custom decorator;

[SO example](https://stackoverflow.com/questions/64497615/how-to-add-a-custom-decorator-to-a-fastapi-route)

Here is one that I created and am using on endpoints where I have a common set of features; 

```
def qmql(handler):
    """Add filter functionality to and endpoint."""

    async def wrapper(
        query: QmqlQuery = Depends(),
        echo: bool = Query(
            default=False, description="If true, the query object will be returned with no command execution."
        ),
    ):
        """Support JSON:API like queries."""
        parser = APIQueryParser(query.raw)
        q = parser.parse()
        echo = getattr(echo, "default", echo)
        if echo:
            parsed = q.json()
            content = EchoResponse(query=query.raw, parsed_query=parsed).dict()
            return JSONResponse(status_code=209, content=content)
        return await handler(q)

    return wrapper
```

**Comment by kauedg** [comment]
> An approach that I've found not to be too awful is to use a custom decorator;

Thanks for your reply. I've considered this approach but I want to have the validation logic in the `Username` class itself.

**Comment by YuriiMotov** [comment, marked as answer]
In python 3.10+ you can create type alias as follows:
```python
UserNamePath: TypeAlias = Annotated[str, Path(..., min_length=3, max_length=8)]
```

and then use it:

```python
from typing import Annotated, TypeAlias

from fastapi import FastAPI, Path

app = FastAPI()

UserNamePath: TypeAlias = Annotated[str, Path(..., min_length=3, max_length=8)]

@app.post(path="/user/{username}")
def create_user(
    *,
    username: UserNamePath,
):
    return {"username": username}

```
